Source/WebCore/rendering/style/RenderStyle.cpp

```diff
@@ -719,6 +719,8 @@ inline bool RenderStyle::changeAffectsVisualOverflow(const RenderStyle& other) c
         // is specified. We can take an early out here.
         auto isVertialWritingMode = isVerticalWritingMode() || other.isVerticalWritingMode();
         auto isAlignedForUnder = textUnderlinePosition() == TextUnderlinePosition::Under || other.textUnderlinePosition() == TextUnderlinePosition::Under
+            || (textUnderlinePosition() == TextUnderlinePosition::Auto && textUnderlineOffset().isAuto())
+            || (other.textUnderlinePosition() == TextUnderlinePosition::Auto && other.textUnderlineOffset().isAuto())
```

Comment:
This is not quite correct as auto should only be considered "under" when in non-horizontal writing mode (see isAlignedForUnder in InlineTextBoxStyle). Ideally, to make this code less error-prone, we would share the logic, something along the lines of:

```
diff --git a/Source/WebCore/rendering/style/RenderStyle.cpp b/Source/WebCore/rendering/style/RenderStyle.cpp
index f8dcc20c1215..f69f99829dc2 100644
--- a/Source/WebCore/rendering/style/RenderStyle.cpp
+++ b/Source/WebCore/rendering/style/RenderStyle.cpp
@@ -717,10 +717,7 @@ inline bool RenderStyle::changeAffectsVisualOverflow(const RenderStyle& other) c
         || m_rareInheritedData->textUnderlinePosition != other.m_rareInheritedData->textUnderlinePosition) {
         // Underlines are always drawn outside of their textbox bounds when text-underline-position: under;
         // is specified. We can take an early out here.
-        auto isVertialWritingMode = isVerticalWritingMode() || other.isVerticalWritingMode();
-        auto isAlignedForUnder = textUnderlinePosition() == TextUnderlinePosition::Under || other.textUnderlinePosition() == TextUnderlinePosition::Under
-            || (isVertialWritingMode && (textUnderlinePosition() == TextUnderlinePosition::Right || other.textUnderlinePosition() == TextUnderlinePosition::Right || textUnderlinePosition() == TextUnderlinePosition::Left || other.textUnderlinePosition() == TextUnderlinePosition::Left));
-        if (isAlignedForUnder)
+        if (isAlignedForUnder(*this) || isAlignedForUnder(other))
             return true;
         return visualOverflowForDecorations(*this) != visualOverflowForDecorations(other);
     }
diff --git a/Source/WebCore/style/InlineTextBoxStyle.cpp b/Source/WebCore/style/InlineTextBoxStyle.cpp
index 8c59b0a974cf..a0b3d8b664d2 100644
--- a/Source/WebCore/style/InlineTextBoxStyle.cpp
+++ b/Source/WebCore/style/InlineTextBoxStyle.cpp
@@ -42,21 +42,6 @@ struct UnderlineOffsetArguments {
     std::optional<TextUnderlinePositionUnder> textUnderlinePositionUnder { };
 };
 
-static bool isAlignedForUnder(const RenderStyle& decoratingBoxStyle)
-{
-    auto underlinePosition = decoratingBoxStyle.textUnderlinePosition();
-    if (underlinePosition == TextUnderlinePosition::Under)
-        return true;
-    if (decoratingBoxStyle.isHorizontalWritingMode())
-        return false;
-    if (underlinePosition == TextUnderlinePosition::Left || underlinePosition == TextUnderlinePosition::Right) {
-        // In vertical typographic modes, the underline is aligned as for under for 'left' and 'right'.
-        return true;
-    }
-    // When left/right support is not enabled.
-    return underlinePosition == TextUnderlinePosition::Auto && decoratingBoxStyle.textUnderlineOffset().isAuto();
-}
-
 static bool isAncestorAndWithinBlock(const RenderInline& ancestor, const RenderObject* child)
 {
     const RenderObject* object = child;
@@ -279,6 +264,21 @@ static GlyphOverflow computedVisualOverflowForDecorations(const RenderStyle& lin
     return overflowResult;
 }
 
+bool isAlignedForUnder(const RenderStyle& decoratingBoxStyle)
+{
+    auto underlinePosition = decoratingBoxStyle.textUnderlinePosition();
+    if (underlinePosition == TextUnderlinePosition::Under)
+        return true;
+    if (decoratingBoxStyle.isHorizontalWritingMode())
+        return false;
+    if (underlinePosition == TextUnderlinePosition::Left || underlinePosition == TextUnderlinePosition::Right) {
+        // In vertical typographic modes, the underline is aligned as for under for 'left' and 'right'.
+        return true;
+    }
+    // When left/right support is not enabled.
+    return underlinePosition == TextUnderlinePosition::Auto && decoratingBoxStyle.textUnderlineOffset().isAuto();
+}
+
 GlyphOverflow visualOverflowForDecorations(const InlineIterator::LineBoxIterator& lineBox, const RenderText& renderer, float textBoxLogicalTop, float textBoxLogicalBottom)
 {
     auto& style = lineBox->isFirst() ? renderer.firstLineStyle() : renderer.style();
diff --git a/Source/WebCore/style/InlineTextBoxStyle.h b/Source/WebCore/style/InlineTextBoxStyle.h
index 0323ac78ff27..e735305394e2 100644
--- a/Source/WebCore/style/InlineTextBoxStyle.h
+++ b/Source/WebCore/style/InlineTextBoxStyle.h
@@ -63,5 +63,6 @@ GlyphOverflow visualOverflowForDecorations(const InlineIterator::LineBoxIterator
 
 float underlineOffsetForTextBoxPainting(const InlineIterator::InlineBox&, const RenderStyle&);
 float overlineOffsetForTextBoxPainting(const InlineIterator::InlineBox&, const RenderStyle&);
+bool isAlignedForUnder(const RenderStyle&);
 
 } // namespace WebCore

```

Reply:
Thanks, that makes sense, I wasn't so sure about this change and also wondered if they can share same criteria.

Reply:
> Thanks, that makes sense, I wasn't so sure which isAlignedForUnder is correct

Relevant CSS spec is to the rescue! :)

Reply:
💯